File: web/src/core/payloads/reaction.ts
/**
 * `reaction.*` payload schemas — the browser port of
 * `server/msgd/core/payloads/reaction.py`.
 *
 * Id fields are *format-validated only*: prefix + ULID validity are checked to
 * catch malformed references early. Referential existence (does the target
 * message exist?) is a server-side concern (§3.2), out of scope here.
 *
 * Emoji domain (LOCKED DECISION, §2.2-style — changing it ⇒ `type_version`
 * bump): `emoji` is a bounded Unicode string, non-empty and at most
 * {@link MAX_EMOJI_BYTES} (64) bytes when UTF-8 encoded. There is deliberately
 * no emoji whitelist — the byte bound is the only gate. Mirrors the Python
 * `_require_emoji` validator so both languages accept/reject the same strings.
 *
 * Idempotency `(message_id, author_user_id, emoji)` (§2.4) is a server
 * projection concern (ENG-97), not enforced here: `author_user_id` lives on the
 * envelope, and these builders validate a single event's shape only.
 */

import { IdKind, isValidTypedId } from '../ids'

/** Upper bound on the UTF-8 byte length of a reaction `emoji` (locked at v1). */
export const MAX_EMOJI_BYTES = 64

const utf8 = new TextEncoder()

/** Throw unless `emoji` is a non-empty string of at most `MAX_EMOJI_BYTES` UTF-8 bytes. */
function requireEmoji(emoji: string): string {
  if (emoji === '') {
    throw new Error('emoji must be non-empty')
  }
  const n = utf8.encode(emoji).length
  if (n > MAX_EMOJI_BYTES) {
    throw new Error(`emoji is ${n} bytes UTF-8, exceeds the ${MAX_EMOJI_BYTES}-byte limit`)
  }
  return emoji
}

function requireMessageId(messageId: string): string {
  if (!isValidTypedId(messageId, IdKind.MESSAGE)) {
    throw new Error(`message_id is not a valid m_ id: ${messageId}`)
  }
  return messageId
}

/** Payload for `reaction.added` v1 (§2.2). */
export type ReactionAddedV1 = {
  message_id: string
  emoji: string
}

/** Payload for `reaction.removed` v1 (§2.2). */
export type ReactionRemovedV1 = {
  message_id: string
  emoji: string
}

/** Options for the reaction payload builders; mirrors the Python model fields. */
export interface BuildReactionPayloadOptions {
  message_id: string
  emoji: string
}

/**
 * Format-validate a `reaction.added` v1 payload.
 *
 * @throws {Error} on a malformed `message_id` or an out-of-domain `emoji`.
 */
export function buildReactionAddedPayload(options: BuildReactionPayloadOptions): ReactionAddedV1 {
  return {
    message_id: requireMessageId(options.message_id),
    emoji: requireEmoji(options.emoji),
  }
}

/**
 * Format-validate a `reaction.removed` v1 payload.
 *
 * @throws {Error} on a malformed `message_id` or an out-of-domain `emoji`.
 */
export function buildReactionRemovedPayload(
  options: BuildReactionPayloadOptions,
): ReactionRemovedV1 {
  return {
    message_id: requireMessageId(options.message_id),
    emoji: requireEmoji(options.emoji),
  }
}

File: server/msgd/core/payloads/reaction.py
"""``reaction.*`` payload schemas (TDD §2.2 / §2.4).

Same modeling discipline as
:class:`~msgd.core.payloads.message.MessageCreatedV1`:

* ``model_config = ConfigDict(extra="allow")`` so additive-only v1 changes
  (§2.3.2) round-trip losslessly through an older reader.
* **Format-validation only** — ``message_id`` prefix + ULID validity and the
  emoji domain (below).  Referential *existence* (does the target message
  exist?) is a server concern (§3.2), never enforced here.

**Emoji domain (LOCKED DECISION, §2.2-style — changing it ⇒ ``type_version``
bump):** ``emoji`` is a bounded **Unicode string**: non-empty and at most
:data:`MAX_EMOJI_BYTES` (64) bytes when UTF-8 encoded.  There is deliberately
**no server-side emoji whitelist** — clients may send any Unicode grapheme
(base emoji, ZWJ sequences, skin-tone modifiers, keycaps, or even a short
non-emoji string); the 64-byte bound is the only gate.  Widening or narrowing
this domain later (e.g. adding a whitelist, or raising the byte cap) is a
breaking change under D9 and must arrive via a ``type_version`` bump, exactly
like ``message.created.format``.

**Idempotency is NOT enforced here.** The reaction set is keyed on
``(message_id, author_user_id, emoji)`` (§2.4) — ``author_user_id`` lives on the
envelope, not the payload, and dedup/idempotent set semantics are a server
projection concern (ENG-97).  This model validates a single reaction event's
shape only.
"""

from __future__ import annotations

from typing import Final

from pydantic import BaseModel, ConfigDict, field_validator

from msgd.core import ids

__all__ = ["MAX_EMOJI_BYTES", "ReactionAddedV1", "ReactionRemovedV1"]

#: Upper bound on the UTF-8 byte length of a reaction ``emoji`` (locked at
#: ``type_version`` 1; see the module docstring).  64 bytes comfortably holds
#: the longest realistic ZWJ emoji sequence while bounding the payload.
MAX_EMOJI_BYTES: Final = 64


def _require_message_id(value: str) -> str:
    if not ids.is_valid_typed_id(value, ids.IdKind.MESSAGE):
        raise ValueError(f"message_id is not a valid m_ id: {value!r}")
    return value


def _require_emoji(value: str) -> str:
    if value == "":
        raise ValueError("emoji must be non-empty")
    n = len(value.encode("utf-8"))
    if n > MAX_EMOJI_BYTES:
        raise ValueError(f"emoji is {n} bytes UTF-8, exceeds the {MAX_EMOJI_BYTES}-byte limit")
    return value


class ReactionAddedV1(BaseModel):
    """Payload for ``reaction.added`` v1 (§2.2).

    Idempotent-add on ``(message_id, author_user_id, emoji)`` at projection time
    (§2.4); this model only validates the event shape (see the module docstring).
    """

    model_config = ConfigDict(extra="allow")

    message_id: str
    emoji: str

    @field_validator("message_id")
    @classmethod
    def _check_message_id(cls, value: str) -> str:
        return _require_message_id(value)

    @field_validator("emoji")
    @classmethod
    def _check_emoji(cls, value: str) -> str:
        return _require_emoji(value)


class ReactionRemovedV1(BaseModel):
    """Payload for ``reaction.removed`` v1 (§2.2) — idempotent remove (§2.4)."""

    model_config = ConfigDict(extra="allow")

    message_id: str
    emoji: str

    @field_validator("message_id")
    @classmethod
    def _check_message_id(cls, value: str) -> str:
        return _require_message_id(value)

    @field_validator("emoji")
    @classmethod
    def _check_emoji(cls, value: str) -> str:
        return _require_emoji(value)
